### ISSUE TITLE
BlazorWebView doesn't download files

### ISSUE THREAD

**Issue by konstantin-karpukhin** [dotnet/maui#15493]
### Description

We have an article about file downloads in Blazor Server or Wasm application [https://learn.microsoft.com/en-us/aspnet/core/blazor/file-downloads?view=aspnetcore-7.0](url).

None of the following approaches works in BlazorWebView for Android/macOS/iOS. Perfectly working in WebView2 on windows.

### Steps to Reproduce

1. Run application
2. Click any link or button for download

Expected behavior:
Files are downloaded on any platform

Actual behavior:
Files are not downloaded on Android/iOS/macOS

### Link to public reproduction project repository

https://github.com/konstantin-karpukhin/BlazorWebViewDownload/tree/main

### Version with bug

7.0.49

### Last version that worked well

Unknown/Other

### Affected platforms

iOS, Android, macOS

### Affected platform versions

all

### Did you find any workaround?

Uncomment '// WORKAROUND' lines in source project.

1. Modify microsoft article code:
add document.body.appendChild(anchorElement);
before anchorElement.click();
2. Download blob/file and convert it to data url using window.onClick in javascript.
3. Call c# code to save file manuall..

It works in case you have access to source code. Didn't find available workaround in case you are using third-party libraries or load external web sites.

**Comment by drasticactions** [comment]
I don't believe this is a MAUI, Blazor, or MAUI Blazor bug. This is the default behavior of the underlying WebViews for these platforms.

https://github.com/drasticactions/MauiRepros/tree/main/BlazorTests

I've created a reproduction project to show this. It's a Blazor WASM app and a series of native .NET apps with WebViews (The same MAUI uses for its webview). Taking your code and running it here, you can see the exact same behavior. Moreover, nearly all of your code is straight JavaScript inside of the index.html. That's unrelated to Blazor or MAUI, you're just using JavaScript, you can reproduce this with straight HTML inside a generic webview with no Blazor at all.

A WebView is not the same as full blown Browser window (Like Safari or Chrome, etc). WebView2 does support baseline downloading support but Android and iOS/Catalyst WKWebView require extra configuration (https://github.com/dotnet/maui/discussions/15322) and that's not in scope for the MAUI WebView at the moment.

@Eilon Does this track?

**Comment by mkArtakMSFT** [comment]
@konstantin-karpukhin because the application runs on the same device already where from does your example tries to download the content? Presumable that content is already on the same device. Can you clarify this please?

If, however, your "download source" is indeed hosted on an external server, it may be due to app configuration / permissions, as pointed out in the above comment by @drasticactions.

**Comment by drasticactions** [comment]
> Our customers encountered an issue with sharing code between different Blazor hosting models. The code works well for Blazor Server/WASM/Hybrid(Windows) applications, but not for non-Windows Blazor MAUI. 

It's not Blazor MAUI. Your issue is fundamental to WebViews on these platforms. If your customers ran your code in an Android, iOS, or Catalyst webview either with .NET or native, they would see the same issue. Likewise, as I said above, your Javascript code has nothing to do with Blazor.

That said, as I pointed to in my reply and as you saw yourself, there are APIs within those native webviews where file links could be handled in some way. You could implement them yourself to handle those usecases right now, and It's not impossible for us to someday create a wrapper around it for MAUI's WebView or BlazorWebView, but that's a feature request.

IMO this is not a bug. The native webviews for iOS/Android don't handle file download links by default, and that behavior is the same between Xamarin.Forms and MAUI, which did not try and create APIs around them. It is possible to do so, but that's a feature.

**Comment by densen2014** [comment]
do somethig like this

```
        _blazorWebView.UrlLoading +=
             (sender, urlLoadingEventArgs) =>
             {
                 if (urlLoadingEventArgs.Url.Host != "0.0.0.0")
                 {
                     //外部链接WebView内打开,例如pdf浏览器
                     Console.WriteLine(urlLoadingEventArgs.Url);
                     urlLoadingEventArgs.UrlLoadingStrategy =
                         UrlLoadingStrategy.OpenInWebView;

                     //拦截可处理 IOS || MACCATALYST 下载文件, 简单测试一下
                     if (urlLoadingEventArgs.Url.ToString().EndsWith(".exe"))
                     {
                         Task.Run(async () => await DownloadAsync(urlLoadingEventArgs.Url));
                     }

                 }
             };
```

**Comment by thomasbach-dk** [comment]
Any news regarding this?

**Comment by Eilon** [comment]
No updates yet, but it might be possible if we implement this feature request: https://github.com/dotnet/maui/issues/11382

**Comment by Zhanglirong-Winnie** [comment]
Verified this issue with Visual Studio Enterprise 17.9.0 Preview 2. Can repro on android platform with sample project.
https://github.com/konstantin-karpukhin/BlazorWebViewDownload/tree/main
